File: apps/api/app/schemas/sale.py
from datetime import datetime
from uuid import UUID
from typing import Optional, List
from decimal import Decimal
from pydantic import BaseModel, ConfigDict
from app.models.sale import SaleStatus


class SaleItemCreate(BaseModel):
    product_id: UUID
    quantity: Decimal
    unit_price: Decimal
    discount_amount: Decimal = Decimal("0.00")


class SalePaymentCreate(BaseModel):
    payment_method: str
    amount: Decimal
    change_amount: Decimal = Decimal("0.00")


class SaleCreate(BaseModel):
    customer_id: Optional[UUID] = None
    items: List[SaleItemCreate]
    payments: List[SalePaymentCreate]
    discount_amount: Decimal = Decimal("0.00")
    notes: Optional[str] = None


class SaleItemResponse(BaseModel):
    id: UUID
    sale_id: UUID
    product_id: UUID
    item_number: int
    product_name: str
    unit_code: str
    quantity: Decimal
    unit_price: Decimal
    discount_amount: Decimal
    total_price: Decimal
    ncm: Optional[str] = None
    cest: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class SalePaymentResponse(BaseModel):
    id: UUID
    sale_id: UUID
    payment_method: str
    amount: Decimal
    change_amount: Decimal

    model_config = ConfigDict(from_attributes=True)


class SaleResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    company_id: UUID
    cash_register_id: Optional[UUID] = None
    customer_id: Optional[UUID] = None
    user_id: UUID
    code: str
    status: SaleStatus
    subtotal: Decimal
    discount_amount: Decimal
    total_amount: Decimal
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SaleDetailResponse(SaleResponse):
    customer_name: Optional[str] = None
    user_name: Optional[str] = None
    items: List[SaleItemResponse] = []
    payments: List[SalePaymentResponse] = []

    model_config = ConfigDict(from_attributes=True)

File: apps/desktop/src/types/sale.ts
export interface CartItem {
  id: string; // Product UUID
  code?: string | null;
  name: string;
  unit_code: string;
  unit_price: number;
  quantity: number;
  discount_amount: number;
  total_price: number;
  ncm?: string | null;
  cest?: string | null;
}

export interface SalePaymentInput {
  payment_method: string;
  amount: number;
  change_amount: number;
}

export interface SaleCreateInput {
  customer_id?: string | null;
  items: {
    product_id: string;
    quantity: number;
    unit_price: number;
    discount_amount: number;
  }[];
  payments: SalePaymentInput[];
  discount_amount: number;
  notes?: string;
}

export interface SaleItem {
  id: string;
  sale_id: string;
  product_id: string;
  item_number: number;
  product_name: string;
  unit_code: string;
  quantity: number;
  unit_price: number;
  discount_amount: number;
  total_price: number;
  ncm?: string | null;
  cest?: string | null;
}

export interface SalePayment {
  id: string;
  sale_id: string;
  payment_method: string;
  amount: number;
  change_amount: number;
}

export interface SaleDetail {
  id: string;
  tenant_id: string;
  company_id: string;
  cash_register_id?: string | null;
  customer_id?: string | null;
  customer_name?: string | null;
  user_id: string;
  user_name?: string | null;
  code: string;
  status: "COMPLETED" | "CANCELED";
  subtotal: number;
  discount_amount: number;
  total_amount: number;
  notes?: string | null;
  items: SaleItem[];
  payments: SalePayment[];
  created_at: string;
  updated_at: string;
}
